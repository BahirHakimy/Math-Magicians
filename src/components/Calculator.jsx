import React from 'react';
import styles from 'styles/Calculator.module.css';
import Button from './Button';
import Display from './Display';
import calculate from '../logic/calculate';

function Calculator() {
  const [result, setResult] = React.useState({
    total: 0,
    next: null,
    operation: null,
  });

  const buttons = [
    'AC',
    '+/-',
    '%',
    'รท',
    '7',
    '8',
    '9',
    'x',
    '4',
    '5',
    '6',
    '-',
    '1',
    '2',
    '3',
    '+',
    '0',
    '.',
    '=',
  ];
  const opreators = ['รท', 'x', '-', '+', '='];
  const { next, total, operation } = result;
  return (
    <div className={styles.container}>
      <h2 className={styles.caption}>Lets do some maths</h2>
      <div className={styles.calculator}>
        <Display
          styles={{ gridColumn: '1/5' }}
          text={`${total || ''} ${operation || ''} ${next || ''}`}
        />
        {buttons.map((text) => (
          <Button
            key={text}
            text={text}
            stylesOverride={{
              gridColumn: text === '0' ? 'span 2' : 'span 1',
              background: opreators.includes(text) ? '#f5913e44' : '#fff4',
              borderBottomLeftRadius: '1rem',
            }}
            onClick={() => setResult(calculate(result, text))}
          />
        ))}
      </div>
    </div>
  );
}

export default Calculator;
