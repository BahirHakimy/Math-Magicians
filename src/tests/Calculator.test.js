import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';
import Calculator from 'components/Calculator';

describe('Tests for Calculator component', () => {
  test('All buttons should be rendered correctly', () => {
    render(<Calculator />);
    expect(screen.getAllByRole('button').length).toBe(19);
  });

  test('The display component inside the Calculator should be rendered correctly', () => {
    render(<Calculator />);
    expect(screen.getByTestId('display')).toBeInTheDocument();
  });

  test('The Calculator should show the results of an operation', async () => {
    render(<Calculator />);
    await userEvent.click(screen.getByText('2'));
    await userEvent.click(screen.getByText('x'));
    await userEvent.click(screen.getByText('5'));
    expect(screen.getByTestId('display')).toHaveTextContent('2 x 5');
    await userEvent.click(screen.getByText('='));
    expect(screen.getByTestId('display')).toHaveTextContent('10');
  });

  test('The Calculator snapshot should be the same each time', () => {
    const element = render(<Calculator />);
    expect(element).toMatchSnapshot();
  });
});
