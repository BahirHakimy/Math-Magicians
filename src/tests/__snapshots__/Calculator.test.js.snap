// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for Calculator component The Calculator snapshot should be the same each time 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <h2
          class="caption"
        >
          Lets do some maths
        </h2>
        <div
          class="calculator"
        >
          <div
            class="display"
            data-testid="display"
          >
              
          </div>
          <button
            class="button"
            style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
            type="button"
          >
            AC
          </button>
          <button
            class="button"
            style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
            type="button"
          >
            +/-
          </button>
          <button
            class="button"
            style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
            type="button"
          >
            %
          </button>
          <button
            class="button"
            style="grid-column: span 1; background: rgba(2, 48, 118, 0.702); border-bottom-left-radius: 1rem;"
            type="button"
          >
            รท
          </button>
          <button
            class="button"
            style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
            type="button"
          >
            7
          </button>
          <button
            class="button"
            style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
            type="button"
          >
            8
          </button>
          <button
            class="button"
            style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
            type="button"
          >
            9
          </button>
          <button
            class="button"
            style="grid-column: span 1; background: rgba(2, 48, 118, 0.702); border-bottom-left-radius: 1rem;"
            type="button"
          >
            x
          </button>
          <button
            class="button"
            style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
            type="button"
          >
            4
          </button>
          <button
            class="button"
            style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
            type="button"
          >
            5
          </button>
          <button
            class="button"
            style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
            type="button"
          >
            6
          </button>
          <button
            class="button"
            style="grid-column: span 1; background: rgba(2, 48, 118, 0.702); border-bottom-left-radius: 1rem;"
            type="button"
          >
            -
          </button>
          <button
            class="button"
            style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
            type="button"
          >
            1
          </button>
          <button
            class="button"
            style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
            type="button"
          >
            2
          </button>
          <button
            class="button"
            style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
            type="button"
          >
            3
          </button>
          <button
            class="button"
            style="grid-column: span 1; background: rgba(2, 48, 118, 0.702); border-bottom-left-radius: 1rem;"
            type="button"
          >
            +
          </button>
          <button
            class="button"
            style="grid-column: span 2; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
            type="button"
          >
            0
          </button>
          <button
            class="button"
            style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
            type="button"
          >
            .
          </button>
          <button
            class="button"
            style="grid-column: span 1; background: rgba(2, 48, 118, 0.702); border-bottom-left-radius: 1rem;"
            type="button"
          >
            =
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <h2
        class="caption"
      >
        Lets do some maths
      </h2>
      <div
        class="calculator"
      >
        <div
          class="display"
          data-testid="display"
        >
            
        </div>
        <button
          class="button"
          style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
          type="button"
        >
          AC
        </button>
        <button
          class="button"
          style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
          type="button"
        >
          +/-
        </button>
        <button
          class="button"
          style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
          type="button"
        >
          %
        </button>
        <button
          class="button"
          style="grid-column: span 1; background: rgba(2, 48, 118, 0.702); border-bottom-left-radius: 1rem;"
          type="button"
        >
          รท
        </button>
        <button
          class="button"
          style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
          type="button"
        >
          7
        </button>
        <button
          class="button"
          style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
          type="button"
        >
          8
        </button>
        <button
          class="button"
          style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
          type="button"
        >
          9
        </button>
        <button
          class="button"
          style="grid-column: span 1; background: rgba(2, 48, 118, 0.702); border-bottom-left-radius: 1rem;"
          type="button"
        >
          x
        </button>
        <button
          class="button"
          style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
          type="button"
        >
          4
        </button>
        <button
          class="button"
          style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
          type="button"
        >
          5
        </button>
        <button
          class="button"
          style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
          type="button"
        >
          6
        </button>
        <button
          class="button"
          style="grid-column: span 1; background: rgba(2, 48, 118, 0.702); border-bottom-left-radius: 1rem;"
          type="button"
        >
          -
        </button>
        <button
          class="button"
          style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
          type="button"
        >
          1
        </button>
        <button
          class="button"
          style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
          type="button"
        >
          2
        </button>
        <button
          class="button"
          style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
          type="button"
        >
          3
        </button>
        <button
          class="button"
          style="grid-column: span 1; background: rgba(2, 48, 118, 0.702); border-bottom-left-radius: 1rem;"
          type="button"
        >
          +
        </button>
        <button
          class="button"
          style="grid-column: span 2; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
          type="button"
        >
          0
        </button>
        <button
          class="button"
          style="grid-column: span 1; background: rgba(34, 34, 34, 0.4); border-bottom-left-radius: 1rem;"
          type="button"
        >
          .
        </button>
        <button
          class="button"
          style="grid-column: span 1; background: rgba(2, 48, 118, 0.702); border-bottom-left-radius: 1rem;"
          type="button"
        >
          =
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
